src/main/groovy/org/shipkit/internal/gradle/GitPlugin.java

```diff
@@ -111,9 +113,16 @@ public void execute(final Task t) {
 
         TaskMaker.execTask(project, COMMIT_CLEANUP_TASK, new Action<Exec>() {
             public void execute(final Exec t) {
-                t.setDescription("Removes last commit, using 'reset --hard HEAD~'");
-                //TODO replace with combination of 'git reset --soft HEAD~ && git stash' so that we don't lose commits
-                t.commandLine("git", "reset", "--hard", "HEAD~");
+                t.setDescription("Stashes current changes");
+                t.commandLine("git", "stash");
+                t.dependsOn(RESET_COMMIT_TASK);
+            }
+        });
+
+        TaskMaker.execTask(project, RESET_COMMIT_TASK, new Action<Exec>() {
+            public void execute(final Exec t) {
```

Comment:
It would be better to create two separate low-level tasks - gitSoftResetCommit and gitStash and one performCommitCleanUp that would depend on them. This is the approach that we use in a few other places, see performGitPush or performRelease. Description of the COMMIT_CLEANUP_TASK is a little confusing and also misleading - it doesn't only stash changes, but also soft-resets them first.